tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py

```diff
@@ -0,0 +1,386 @@
+import pytest
+import numpy as np
+from tardis import constants as c
+from astropy import units as u
+
+from copy import deepcopy
+import numpy.testing as ntest
+from numba import cuda
+from numba import njit
+
+
+import tardis.montecarlo.montecarlo_numba.formal_integral_cuda as formal_integral_cuda
+import tardis.montecarlo.montecarlo_numba.formal_integral as formal_integral_numba
+from tardis.montecarlo.montecarlo_numba.numba_interface import NumbaModel
+
+
+from tardis.montecarlo.montecarlo_numba.formal_integral import FormalIntegrator
+from tardis.montecarlo.montecarlo_numba.formal_integral import (
+    NumbaFormalIntegrator,
+)
+
+from tardis.montecarlo import MontecarloRunner
+
+
+# Test cases must also take into account use of a GPU to run. If there is no GPU then the test cases will fail.
+GPUs_available = cuda.is_available()
+
+
+@pytest.mark.skipif(
+    not GPUs_available, reason="No GPU is available to test CUDA function"
+)
+@pytest.mark.parametrize(
+    ["nu", "T"],
+    [
+        (1e14, 1e4),
+        (0, 1),
+        (1, 1),
+    ],
+)
+def test_intensity_black_body_cuda(nu, T):
+    """
+    Initializes the test of the cuda version
+    against the numba implementation of the
+    intensity_black_body to 15 decimals. This
+    is done as both results have 15 digits of precision.
+    """
+    actual = np.zeros(3)
+    black_body_caller[1, 3](nu, T, actual)
+
+    expected = formal_integral_numba.intensity_black_body(nu, T)
+
+    ntest.assert_allclose(actual, expected, rtol=1e-14)
+
+
+@cuda.jit
+def black_body_caller(nu, T, actual):
+    """
+    This calls the CUDA function and fills out
+    the array
+    """
+    x = cuda.grid(1)
+    actual[x] = formal_integral_cuda.intensity_black_body_cuda(nu, T)
+
+
+@pytest.mark.skipif(
+    not GPUs_available, reason="No GPU is available to test CUDA function"
+)
+@pytest.mark.parametrize(
+    ["N", "N_loc"], [(1e2, 0), (1e3, 1), (1e4, 2), (1e5, 3)]
+)
+def test_trapezoid_integration_cuda(N, N_loc):
+    """
+    Initializes the test of the cuda version
+    against the numba implementation of the
+    trapezoid_integration to 15 decimals. This
+    is done as both results have 15 digits of precision.
+    """
+    actual = np.zeros(4)
+
+    h = 1.0
+    N = int(N)
+    data = np.random.random(N)
+
+    expected = formal_integral_numba.trapezoid_integration(data, h)
+    trapezoid_integration_caller[1, 4](data, h, actual)
+
+    ntest.assert_allclose(actual[N_loc], expected, rtol=1e-14)
+
+
+@cuda.jit
+def trapezoid_integration_caller(data, h, actual):
+    """
+    This calls the CUDA function and fills out
+    the array
+    """
+    x = cuda.grid(1)
+    actual[x] = formal_integral_cuda.trapezoid_integration_cuda(data, h)
+
+
+@njit(fastmath=True, error_model="numpy", parallel=False)
+def calculate_z(r, p, inv_t):
```

Comment:
Why is this here and not in another file?

Reply:
I didn't want to add an import, so I just typed out the header.